File: backend/internal/service/auth_service.go
package service

import (
	"context"
	"errors"
	"time"

	"topsus-krs/internal/domain"
	"topsus-krs/internal/dto"
	"topsus-krs/internal/repository"
	"topsus-krs/utils"
)

// AuthService mendefinisikan kontrak business logic untuk autentikasi
type AuthService interface {
	Register(ctx context.Context, req *dto.RegisterRequest) (*dto.AuthResponse, error)
	Login(ctx context.Context, req *dto.LoginRequest) (*dto.AuthResponse, error)
	GuestLogin(ctx context.Context) (*dto.AuthResponse, error)
	GetProfile(ctx context.Context, userID uint64) (*dto.UserProfile, error)
	UpdateProfile(ctx context.Context, userID uint64, req *dto.UpdateProfileRequest) (*dto.UserProfile, error)
	ChangePassword(ctx context.Context, userID uint64, req *dto.ChangePasswordRequest) error
}

type authService struct {
	userRepo   repository.UserRepository
	jwtSecret  string
	jwtExpiry  int
}

func NewAuthService(userRepo repository.UserRepository, jwtSecret string, jwtExpiry int) AuthService {
	return &authService{
		userRepo:  userRepo,
		jwtSecret: jwtSecret,
		jwtExpiry: jwtExpiry,
	}
}

func (s *authService) Register(ctx context.Context, req *dto.RegisterRequest) (*dto.AuthResponse, error) {
	existing, err := s.userRepo.FindByEmail(ctx, req.Email)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, errors.New("email sudah terdaftar")
	}

	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return nil, err
	}

	user := &domain.User{
		Name:     req.Name,
		Email:    req.Email,
		Password: hashedPassword,
		Role:     req.Role,
	}

	if err := s.userRepo.Create(ctx, user); err != nil {
		return nil, err
	}

	token, err := utils.GenerateToken(user.ID, user.Email, user.Role, user.Name, s.jwtSecret, s.jwtExpiry)
	if err != nil {
		return nil, err
	}

	return &dto.AuthResponse{
		Token: token,
		User: dto.UserProfile{
			ID:        user.ID,
			Name:      user.Name,
			Email:     user.Email,
			Role:      user.Role,
			CreatedAt: user.CreatedAt,
			UpdatedAt: user.UpdatedAt,
		},
	}, nil
}

func (s *authService) Login(ctx context.Context, req *dto.LoginRequest) (*dto.AuthResponse, error) {
	user, err := s.userRepo.FindByEmail(ctx, req.Email)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, errors.New("email atau password salah")
	}

	if !utils.CheckPassword(req.Password, user.Password) {
		return nil, errors.New("email atau password salah")
	}

	token, err := utils.GenerateToken(user.ID, user.Email, user.Role, user.Name, s.jwtSecret, s.jwtExpiry)
	if err != nil {
		return nil, err
	}

	return &dto.AuthResponse{
		Token: token,
		User: dto.UserProfile{
			ID:        user.ID,
			Name:      user.Name,
			Email:     user.Email,
			Role:      user.Role,
			CreatedAt: user.CreatedAt,
			UpdatedAt: user.UpdatedAt,
		},
	}, nil
}

func (s *authService) GuestLogin(ctx context.Context) (*dto.AuthResponse, error) {
	token, err := utils.GenerateToken(0, "guest@krs.ac.id", "guest", "Tamu", s.jwtSecret, 2)
	if err != nil {
		return nil, err
	}
	now := time.Now()
	return &dto.AuthResponse{
		Token: token,
		User: dto.UserProfile{
			ID:        0,
			Name:      "Tamu",
			Email:     "guest@krs.ac.id",
			Role:      "guest",
			CreatedAt: now,
			UpdatedAt: now,
		},
	}, nil
}

func (s *authService) GetProfile(ctx context.Context, userID uint64) (*dto.UserProfile, error) {
	user, err := s.userRepo.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, errors.New("user tidak ditemukan")
	}

	return &dto.UserProfile{
		ID:        user.ID,
		Name:      user.Name,
		Email:     user.Email,
		Role:      user.Role,
		CreatedAt: user.CreatedAt,
		UpdatedAt: user.UpdatedAt,
	}, nil
}

func (s *authService) UpdateProfile(ctx context.Context, userID uint64, req *dto.UpdateProfileRequest) (*dto.UserProfile, error) {
	user, err := s.userRepo.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, errors.New("user tidak ditemukan")
	}

	user.Name = req.Name
	if err := s.userRepo.Update(ctx, user); err != nil {
		return nil, err
	}

	return &dto.UserProfile{
		ID:        user.ID,
		Name:      user.Name,
		Email:     user.Email,
		Role:      user.Role,
		CreatedAt: user.CreatedAt,
		UpdatedAt: user.UpdatedAt,
	}, nil
}

func (s *authService) ChangePassword(ctx context.Context, userID uint64, req *dto.ChangePasswordRequest) error {
	user, err := s.userRepo.FindByID(ctx, userID)
	if err != nil {
		return err
	}
	if user == nil {
		return errors.New("user tidak ditemukan")
	}

	if !utils.CheckPassword(req.OldPassword, user.Password) {
		return errors.New("password lama tidak sesuai")
	}

	newHash, err := utils.HashPassword(req.NewPassword)
	if err != nil {
		return err
	}

	user.Password = newHash
	return s.userRepo.Update(ctx, user)
}

File: backend-refactor-python/app/services/auth_service.py
from datetime import datetime, timezone

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from ..repositories.user_repository import UserRepository
from ..schemas.auth import (
    AuthResponse,
    ChangePasswordRequest,
    LoginRequest,
    RegisterRequest,
    UpdateProfileRequest,
    UserProfile,
)
from ..security import create_token, hash_password, verify_password
from ..config import settings


class AuthService:
    def __init__(self, db: Session):
        self.repo = UserRepository(db)

    def register(self, req: RegisterRequest) -> AuthResponse:
        if self.repo.find_by_email(req.email):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Email sudah terdaftar",
            )
        hashed = hash_password(req.password)
        user = self.repo.create(req.name, req.email, hashed, req.role)
        token = create_token(user.id, user.email, user.role, user.name)
        return AuthResponse(
            token=token,
            user=UserProfile(
                id=user.id,
                name=user.name,
                email=user.email,
                role=user.role,
                created_at=user.created_at,
                updated_at=user.updated_at,
            ),
        )

    def login(self, req: LoginRequest) -> AuthResponse:
        user = self.repo.find_by_email(req.email)
        if not user or not verify_password(req.password, user.password):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Email atau password salah",
            )
        token = create_token(user.id, user.email, user.role, user.name)
        return AuthResponse(
            token=token,
            user=UserProfile(
                id=user.id,
                name=user.name,
                email=user.email,
                role=user.role,
                created_at=user.created_at,
                updated_at=user.updated_at,
            ),
        )

    @staticmethod
    def guest_login() -> AuthResponse:
        token = create_token(0, "guest@krs.ac.id", "guest", "Tamu", expiry_hours=2)
        now = datetime.now(timezone.utc)
        return AuthResponse(
            token=token,
            user=UserProfile(
                id=0,
                name="Tamu",
                email="guest@krs.ac.id",
                role="guest",
                created_at=now,
                updated_at=now,
            ),
        )

    def get_profile(self, user_id: int) -> UserProfile:
        user = self.repo.find_by_id(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="User tidak ditemukan")
        return UserProfile(
            id=user.id,
            name=user.name,
            email=user.email,
            role=user.role,
            created_at=user.created_at,
            updated_at=user.updated_at,
        )

    def update_profile(self, user_id: int, req: UpdateProfileRequest) -> UserProfile:
        user = self.repo.find_by_id(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="User tidak ditemukan")
        self.repo.update_name(user_id, req.name)
        updated = self.repo.find_by_id(user_id)
        return UserProfile(
            id=updated.id,
            name=updated.name,
            email=updated.email,
            role=updated.role,
            created_at=updated.created_at,
            updated_at=updated.updated_at,
        )

    def change_password(self, user_id: int, req: ChangePasswordRequest) -> None:
        user = self.repo.find_by_id(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="User tidak ditemukan")
        if not verify_password(req.old_password, user.password):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Password lama tidak sesuai",
            )
        self.repo.update_password(user_id, hash_password(req.new_password))
